sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -148,3 +148,110 @@ private[joins] object HashedRelation {
     }
   }
 }
+
+
+/**
+ * A HashedRelation for UnsafeRow, which is backed by BytesToBytesMap that maps the key into a
+ * sequence of values.
+ *
+ * TODO(davies): use BytesToBytesMap
+ */
+private[joins] final class UnsafeHashedRelation(
+    private var hashTable: JavaHashMap[UnsafeRow, CompactBuffer[UnsafeRow]],
+    private var keyTypes: Array[DataType],
+    private var rowTypes: Array[DataType])
+  extends HashedRelation with Externalizable {
+
+  def this() = this(null, null, null)  // Needed for serialization
+
+  // UnsafeProjection is not thread safe
```

Comment:
Can you elaborate on why instances of UnsafeProjection are not thread-safe?  I don't see any mention of this in the Scaladoc for `UnsafeProjection`, so we should probably update it to make any thread-safety concerns clearer.

Does `UnsafeHashedRelation` have to be thread-safe?  I thought it was only used in the context of a single task.